feat: add ability to Reset to BIOS

nmgaston opened the pull request:
Implements boot source override functionality in the Redfish ComputerSystem.Reset action, allowing clients to configure boot settings (e.g., boot to BIOS setup) concurrently with system reset operations. This provides a single atomic operation for power management with boot override, eliminating the need for separate PATCH and reset requests.

**Curl Command**

_First set it up to Reset to BIOS on next reboot_
curl -k -X PATCH https://localhost:8181/redfish/v1/Systems/{id} \
  -u standalone:G@ppm0ym \
  -H 'Content-Type: application/json' \
  -d '{
    "Boot": {
      "BootSourceOverrideTarget": "BiosSetup",
      "BootSourceOverrideEnabled": "Once"
    }
  }'

_Now Reboot the system_
curl -k -X POST https://localhost:8181/redfish/v1/Systems/{id}/Actions/ComputerSystem.Reset \
  -u standalone:G@ppm0ym \
  -H 'Content-Type: application/json' \
  -d '{"ResetType": "ForceRestart"}'

sudhir-intc commented on redfish/openapi/dmtf/ComputerSystem.v1_26_0.yaml:
```diff
@@ -23,7 +23,12 @@ components:
         '@odata.type':
-          $ref: http://redfish.dmtf.org/schemas/v1/odata-v4.yaml#/components/schemas/odata-v4_type
+          $ref: odata-v4.yaml#/components/schemas/odata-v4_type
+        Boot:
```
Boot should be placed after Actions. Alphabetical order to be maintained for ease of comparison with the source file

nmgaston replied:
done.

sudhir-intc commented on redfish/openapi/dmtf/openapi-reduced.yaml:
```diff
@@ -108,6 +108,39 @@ paths:
+            schema:
+              $ref: ComputerSystem.v1_26_0.yaml#/components/schemas/ComputerSystem_v1_26_0_ComputerSystem
+        required: true
+      responses:
```
Not matching the specification. Spec has only description for 200, 204 and default. 
Can you check and update.

nmgaston replied:
I'm not even sure why this was added.  I don't believe Patch is needed at this time.  I'll remove this change.

sudhir-intc commented on redfish/openapi/merged/redfish-openapi.yaml:
```diff
@@ -462,6 +593,11 @@ components:
           $ref: '#/components/schemas/odata-v4_id'
         '@odata.type':
           $ref: '#/components/schemas/odata-v4_type'
+        Boot:
```
Ordering should match the spec, Boot should come after Actions

nmgaston replied:
Done.

sudhir-intc commented on redfish/openapi/dmtf/ComputerSystem.v1_26_0.yaml:
Schema ordering and content not inline with the DMTF 2025.3 version. Please chekc

nmgaston replied:
done.

sudhir-intc commented on internal/usecase/devices/boot.go:
```diff
@@ -0,0 +1,90 @@
+package devices
```
why is this implementation added outside the redfish folder. Should this not be part of the redfish folder.

nmgaston replied:
It needs access to the private fields (uc.repo and uc.device) in the UseCase struct in the main section  If I moved it to the redfish directory, there would still need to be some code in the internal/usecase/devices/ directory to access and relay the data to/from the private fields in the UseCase struct since they are private.

sudhir-intc commented on redfish/openapi/dmtf/ComputerSystem.v1_26_0.yaml:
```diff
@@ -124,6 +129,138 @@ components:
       x-patternProperties:
         ^([a-zA-Z_][a-zA-Z0-9_]*)?@(odata|Redfish|Message)\.[a-zA-Z_][a-zA-Z0-9_]*$:
           description: This property shall specify a valid odata or Redfish property.
+    ComputerSystem_v1_26_0_Boot:
```
ComputerSystem_v1_26_0_Boot
ComputerSystem_v1_26_0_BootSourceOverrideEnabled
ComputerSystem_v1_26_0_BootSourceOverrideMode

Should be placed before   ComputerSystem_v1_26_0_ComputerSystem to maintain the sorted order

sudhir-intc replied:
Additionally there are some description related differences. Please check this https://github.com/DMTF/Redfish-Publications/blob/2025.3/openapi/ComputerSystem.v1_26_0.yaml

nmgaston replied:
Hopefully it's correct now.  I trusted the AI to make the changes and obviously it didn't do very well.

sudhir-intc commented on redfish/openapi/dmtf/ComputerSystem.v1_26_0.yaml:
```diff
@@ -124,6 +129,138 @@ components:
+        Legacy: The system boots in non-UEFI boot mode to the boot source override
+          target.
+        UEFI: The system boots in UEFI boot mode to the boot source override target.
+    ComputerSystem_v1_26_0_BootSourceOverrideTarget:
```
From where this field has been picked as its not part of the DMTF 2025.3 spec. 
Please check this : https://github.com/DMTF/Redfish-Publications/blob/2025.3/openapi/ComputerSystem.v1_26_0.yaml

nmgaston replied:
same as above.  The AI seems to have added something.

sudhir-intc commented on redfish/openapi/dmtf/ComputerSystem.v1_26_0.yaml:
```diff
@@ -11,6 +11,343 @@ components:
+          persistent boot order.
+        BootOrder: The system uses the `BootOrder` property to specify the persistent
+          boot order.
+    ComputerSystem_v1_26_0_BootSource:
```
This schema is not defined in the specification. could you please check.

nmgaston replied:
correct.  The AI had done that to avoid pulling in the ComputerSystem.yaml file.  I had it pull in the file instead.

sudhir-intc commented on redfish/openapi/dmtf/ComputerSystem.v1_26_0.yaml:
```diff
@@ -11,6 +11,343 @@ components:
+            boot order associated with this computer system.
+          items:
+            oneOf:
+            - $ref: '#/components/schemas/ComputerSystem_v1_26_0_BootSource'
```
Value defined in the spec is "          - $ref: http://redfish.dmtf.org/schemas/v1/ComputerSystem.yaml#/components/schemas/ComputerSystem_BootSource" and there's no such definition for ComputerSystem_v1_26_0_BootSource

nmgaston replied:
fixed.

sudhir-intc commented on redfish/openapi/dmtf/ComputerSystem.v1_26_0.yaml:
```diff
@@ -11,6 +11,343 @@ components:
+          description: The BIOS boot mode to use when the system boots from the `BootSourceOverrideTarget`
+            boot source.
+          oneOf:
+          - $ref: '#/components/schemas/ComputerSystem_v1_26_0_BootSourceOverrideMode'
```
The reference defined in the specification is http://redfish.dmtf.org/schemas/v1/ComputerSystem.yaml#/components/schemas/ComputerSystem_BootSource

nmgaston replied:
This is what I see at this point: (I see that link for BootSourceOverrideTarget, but not for BootSourceOverrideMode)

BootSourceOverrideMode:
          description: The BIOS boot mode to use when the system boots from the `BootSourceOverrideTarget`
            boot source.
          oneOf:
          - $ref: '#/components/schemas/ComputerSystem_v1_26_0_BootSourceOverrideMode'
          - enum:
            - null
          readOnly: false
          x-longDescription: This property shall contain the BIOS boot mode to use
            when the system boots from the `BootSourceOverrideTarget` boot source.
          x-versionAdded: v1_1_0

sudhir-intc commented on redfish/openapi/dmtf/ComputerSystem.v1_26_0.yaml:
```diff
@@ -11,6 +11,343 @@ components:
+        Once: On its next boot cycle, the system boots one time to the boot source
+          override target.  Then, the `BootSourceOverrideEnabled` value is reset to
+          `Disabled`.
+    ComputerSystem_v1_26_0_BootSourceOverrideMode:
```
This schema is not defined in the specification

nmgaston replied:
I see it at line 430: https://github.com/DMTF/Redfish-Publications/blob/2025.3/openapi/ComputerSystem.v1_26_0.yaml

 ComputerSystem_v1_26_0_BootSourceOverrideMode:
      enum:
      - Legacy
      - UEFI
      type: string
      x-enumDescriptions:
        Legacy: The system boots in non-UEFI boot mode to the boot source override
          target.
        UEFI: The system boots in UEFI boot mode to the boot source override target.

sudhir-intc commented on redfish/openapi/dmtf/ComputerSystem.v1_26_0.yaml:
```diff
@@ -355,6 +697,33 @@ components:
         DPU: v1_16_0
       x-enumVersionDeprecated:
         OS: v1_21_0
+    ComputerSystem_v1_26_0_StopBootOnFault:
```
This schema is not defined in the spec. Please check

nmgaston replied:
I see it around line 2329 in the spec:  https://github.com/DMTF/Redfish-Publications/blob/2025.3/openapi/ComputerSystem.v1_26_0.yaml

ComputerSystem_v1_26_0_StopBootOnFault:
      enum:
      - Never
      - AnyFault
      type: string
      x-enumDescriptions:
        AnyFault: The system should stop the boot on any fault.
        Never: The system performs any normal recovery actions during boot if a fault
          occurs.
      x-enumLongDescriptions:
        AnyFault: This value shall indicate the system will stop the boot if a fault
          occurs.  This includes, but is not limited to, faults that affect performance,
          fault tolerance, or capacity.
        Never: This value shall indicate the system will continue to attempt to boot
          if a fault occurs.

nmgaston committed:
feat(redfish): Add boot override parameters to Reset action

- Added BootSourceOverrideEnabled, BootSourceOverrideMode, and BootSourceOverrideTarget parameters to ComputerSystem.Reset action
- Updated OpenAPI spec (ComputerSystem.v1_26_0.yaml) to include boot parameters in ResetRequestBody
- Regenerated types, server, and spec files with new boot parameter union types
- Updated systems_actions handler to:
  * Convert request boot parameters from union types to ComputerSystemBoot types
  * Call UpdateBootSettings before executing reset
  * Handle boot settings validation and errors
- Removed debug logging and kept production-level logs
- Successfully tested reset-to-BIOS functionality on Intel NUC with Intel AMT

Allows setting boot configuration (e.g., boot to BIOS setup) concurrently
with system reset action, providing a single atomic operation for power
management with boot override.

Tested-on: Intel NUC9V7QNX with Intel AMT

chore: fix lint issues

chore: fix lint issues

feat(redfish): add boot override parameters to Reset action

- Add boot override parameters to Reset action (BootSourceOverrideTarget, BootSourceOverrideEnabled, BootSourceOverrideMode)
- Create base ComputerSystem.yaml with BootSource enum definitions per DMTF spec
- Update ComputerSystem.v1_26_0.yaml to reference base schema instead of inline definitions
- Fix nlreturn linting errors (add blank lines before returns)
- Remove unused nolint directive
- Regenerate OpenAPI types with proper schema structure

fix: change spec file and regenerate

nmgaston committed:
fix: reboot using standard Redfish method instead of vendor extensions

sudhir-intc reviewed (approved):
LGTM

nmgaston committed:
chore: update golang version for CVE

nmgaston committed:
Merge branch 'main' into AddCyleToBIOSCapability
